islandora_compound_object.module

```diff
@@ -268,41 +268,110 @@ function islandora_compound_object_access(AbstractObject $object) {
+    if ($child == NULL) {
+      $parts = islandora_compound_object_get_parts($parent->id);
+      if (!empty($parts)) {
+        $child = islandora_object_load($parts[0]);
+      }
     }
-    $mime_detector = new MimeDetect();
-    $ext = $mime_detector->getExtension($child['TN']->mimeType);
-    // Windows will likely store temp data in a temp directory
-    // rather than in memory. Since the colon is forbidden in
-    // filenames, replace it with an underscore instead.
-    $thumbnail_id = ((islandora_deployed_on_windows()) ? str_replace(':', '_', $child->id) : $child->id);
-    $file = drupal_realpath("temporary://{$thumbnail_id}_TN.{$ext}");
-    $child['TN']->getContent($file);
-    if (empty($parent['TN'])) {
-      $ds = $parent->constructDatastream('TN', 'M');
-      $parent->ingestDatastream($ds);
+    if ($child != NULL && !empty($child['TN'])) {
+      $mime_detector = new MimeDetect();
+      $ext = $mime_detector->getExtension($child['TN']->mimeType);
+      // Windows will likely store temp data in a temp directory
+      // rather than in memory. Since the colon is forbidden in
+      // filenames, replace it with an underscore instead.
+      $thumbnail_id = ((islandora_deployed_on_windows()) ? str_replace(':', '_', $child->id) : $child->id);
+      $file = drupal_realpath("temporary://{$thumbnail_id}_TN.{$ext}");
+      $child['TN']->getContent($file);
+      if (empty($parent['TN'])) {
+        $ds = $parent->constructDatastream('TN', 'M');
+        $ds->setContentFromFile($file, TRUE);
+        $ds->label = $child['TN']->label;
+        $ds->mimetype = $child['TN']->mimetype;
+        $parent->ingestDatastream($ds);
+      }
+      else {
+        // Check the size and then the content for equality.
+        // Only set if both are unequal.
+        if (($parent['TN']->size != $child['TN']->size) &&
+          $parent['TN']->content != $child['TN']->content) {
+          // Check label equality before setting label.
+          if ($parent['TN']->label != $child['TN']->label) {
+            $parent['TN']->label = $child['TN']->label;
+          }
+          // Check mimetype equality before setting mimetype.
+          if ($parent['TN']->mimeType != $child['TN']->mimetype) {
+            $parent['TN']->mimetype = $child['TN']->mimetype;
+          }
+          $parent['TN']->setContentFromFile($file, TRUE);
+        }
+      }
+      file_unmanaged_delete($file);
     }
-    else {
-      $ds = $parent['TN'];
+    // Child doesn't have a thumbnail then we can't set the parents TN either.
+    elseif (isset($parent['TN'])) {
+      $parent->purgeDatastream('TN');
     }
-    $ds->setContentFromFile($file, TRUE);
-    $ds->label = 'Thumbnail';
-    $ds->mimetype = $mime_detector->getMimetype($file);
-
-    file_unmanaged_delete($file);
   }
-  elseif (isset($parent['TN'])) {
-    $parent->purgeDatastream('TN');
+  catch (Exception $e) {
+    throw ($e);
```

Comment:
actually why catch and then throw immediately? would be fine to just let 'er on through?